README.md
<!--
This README describes the package. If you publish this package to pub.dev,
this README's contents appear on the landing page for your package.

For information about how to write a good package README, see the guide for
[writing package pages](https://dart.dev/guides/libraries/writing-package-pages).

For general information about developing packages, see the Dart guide for
[creating packages](https://dart.dev/guides/libraries/create-library-packages)
and the Flutter guide for
[developing packages and plugins](https://flutter.dev/developing-packages).
-->


# Responsive Text Widget

```responsive_text_widget```  Flutter package to make your texts Responsive. The Package will Assist you to calculate size of texts. Prevents text overflow without explicitly setting a fixed font size. ✨

It's support any type of texts you want to show such as [ headline, bodytext, subtitle text or caption text], etc.


## Installation

Create a new project with the command

```yaml
flutter create MyApp
```

Add

```yaml
responsive_text_widget: ...
```

to your `pubspec.yaml` of your flutter project.
**OR**
run

```yaml
flutter pub add responsive_text_widget
```

in your project's root directory.

In your library add the following import:

```dart
import 'package:responsive_text_widget/responsive_text_widget.dart';
```

For help getting started with Flutter, view the online [documentation](https://flutter.io/).

## Usage

```dart
class MyApp extends StatelessWidget {
  const MyApp({super.key});

  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Responsive Text',
      theme: ThemeData(
        colorScheme: ColorScheme.fromSeed(seedColor: Colors.deepPurple),
        useMaterial3: true,
      ),
      home: const MyHomePage(),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({super.key});

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: const Text("Responsive Text Widget"),
      ),
      body: Padding(
        padding: const EdgeInsets.all(16),
        child: Center(
          child: Column(
            mainAxisAlignment: MainAxisAlignment.center,
            children: <Widget>[
              const Row(
                children: [
                  Expanded(
                    child: ResponsiveText(
                      text: "Responsive Text",
                      style: TextStyle(
                        color: Colors.black,
                        fontWeight: FontWeight.bold,
                        fontSize: 15,
                        fontFamily: "open sans",
                      ),
                    ),
                  ),
                  Expanded(
                    child: ResponsiveText(
                      text:
                          "Flutter package to make your texts Responsive. The Package will Assist you to calculate size of texts.",
                      style: TextStyle(
                        color: Colors.black,
                        fontWeight: FontWeight.bold,
                        fontSize: 15,
                      ),
                    ),
                  )
                ],
              ),
              const SizedBox(height: 30),
              TextButton(
                child: const ResponsiveText(
                  text: "Text Button",
                  textAlign: TextAlign.center,
                  style: TextStyle(
                    color: Colors.orange,
                    fontWeight: FontWeight.bold,
                    fontSize: 15,
                  ),
                ),
                onPressed: () async {},
              )
            ],
          ),
        ),
      ),
    );
  }
}
```


## Constructor

#### Basic

| Parameter    | Default                                                                                          | Description                                        | Required |
|--------------|:-------------------------------------------------------------------------------------------------|:---------------------------------------------------|:--------:|
| text         | -                                                                                                | Text widget required text                          |   True   |
| textStyle    | const TextStyle(color: Colors.black)                                                             | Style properties for text style                    |  false   |
| textAlign    | TextAlign.center                                                                                 | Align the text on the left right, center, top edge |  false   |
| textOverflow | TextOverflow.visible                                                                             | Clip the overflowing text to fix its container.    |  false   |


CHANGELOG.md
## 0.0.1

- 0.0.1: initial release of responsive text widget package..


test/responsive_text_widget_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:responsive_text_widget/responsive_text_widget.dart';

void main() {
  test('adds one to input values', () {
    // final calculator = Calculator();
    // expect(calculator.addOne(2), 3);
    // expect(calculator.addOne(-7), -6);
    // expect(calculator.addOne(0), 1);
  });
}


lib/responsive_text_widget.dart
/// responsive text widget library class.
library responsive_text_widget;

import 'package:flutter/material.dart';

/// A Responsive Text Class.
class ResponsiveText extends StatelessWidget {
  final String? text;
  final TextStyle? style;
  final TextAlign textAlign;
  final int maxLine;
  final TextOverflow overflow;

  const ResponsiveText({
    Key? key,
    required this.text,
    this.style = const TextStyle(color: Colors.black),
    this.textAlign = TextAlign.center,
    this.maxLine = 100,
    this.overflow = TextOverflow.visible,
  }) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return RichText(
      text: TextSpan(
        text: text,
        style: style,
      ),
      maxLines: maxLine,
      overflow: overflow,
      textAlign: textAlign,
    );
  }
}

/// Calculates the font size to fit the available space.
///
/// This method calculates and returns the adjusted font size based on the
/// [TextPainter] measurements and the maximum available space for both width
/// and height.
/*
double calculateFontSizeToFit(
    TextPainter textPainter,
    double maxWidth,
    double maxHeight,
    double currentFontSize,
    ) {
  final double scaleFactor2 = currentFontSize / textPainter.width;
  if (textPainter.width > maxWidth) {
    currentFontSize = maxWidth * scaleFactor2;
  }

  final TextPainter textPainter2 = TextPainter(
    text: TextSpan(
      text: child.data,
      style: child.style?.copyWith(fontSize: currentFontSize),
    ),
    textDirection: TextDirection.ltr,
    maxLines: child.maxLines ?? 1,
  );
  textPainter2.layout(maxWidth: double.maxFinite);

  if (textPainter2.height > maxHeight) {
    final double scaleFactor = currentFontSize / textPainter.height;
    currentFontSize = maxHeight * scaleFactor;
  }
*/


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:responsive_text_widget/responsive_text_widget.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Responsive Text',
      theme: ThemeData(
        colorScheme: ColorScheme.fromSeed(seedColor: Colors.deepPurple),
        useMaterial3: true,
      ),
      home: const MyHomePage(),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({super.key});

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: const Text("Responsive Text Widget"),
      ),
      body: Padding(
        padding: const EdgeInsets.all(16),
        child: Center(
          child: Column(
            mainAxisAlignment: MainAxisAlignment.center,
            children: <Widget>[
              const Row(
                children: [
                  Expanded(
                    child: ResponsiveText(
                      text: "Responsive Text",
                      style: TextStyle(
                        color: Colors.black,
                        fontWeight: FontWeight.bold,
                        fontSize: 15,
                        fontFamily: "open sans",
                      ),
                    ),
                  ),
                  Expanded(
                    child: ResponsiveText(
                      text:
                          "Flutter package to make your texts Responsive. The Package will Assist you to calculate size of texts.",
                      style: TextStyle(
                        color: Colors.black,
                        fontWeight: FontWeight.bold,
                        fontSize: 15,
                      ),
                    ),
                  )
                ],
              ),
              const SizedBox(height: 30),
              TextButton(
                child: const ResponsiveText(
                  text: "Text Button",
                  textAlign: TextAlign.center,
                  style: TextStyle(
                    color: Colors.orange,
                    fontWeight: FontWeight.bold,
                    fontSize: 15,
                  ),
                ),
                onPressed: () async {},
              )
            ],
          ),
        ),
      ),
    );
  }
}
